### DISCUSSION TITLE
[@types/react] forwardRef problem on type validation for discriminating unions

### DISCUSSION THREAD

**Discussion by Mintoo200** [DefinitelyTyped/DefinitelyTyped#67428] [Issues with a @types package]
It would seem that forwarding a ref causes type validation problems in some cases when the component's props include a discriminating union.
I made a [simple example on CodeSandbox](https://codesandbox.io/s/failed-type-validation-on-forwarded-ref-v9nj2d) and on [Typescript playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgcilQ3wG4AocmATzCTgAUcwBnOAXnLjgB84BvLt2EAiFFGAoAtABsUAIyQyRALjgsYEgHYBzCsNHjJshUplIAJvOoiA-Go3a9Q7gF8hfQQe5iJ0uYrK9uqawLr63r7GAWaW1qohThFuFOSYAK5aGMAQWnAAstQAwriQWkhaMADqwDAAFhDpMMiYABRgzCxqTBCsAJQCQkQw6VB5ADwWwABuAgB0Cx29LK4AfAAqSBrjAPRT06sU7uRouRoFxaW5FdW1dS0ciMQwc5jQAO7iFi3jABLr+QAMgARGYAUXMIBuABpGJ1Vq0hBksjAcnlCiVwNdKjV6i12p1YURMAMvMJhqM8ojvHBJjM4MT2Pxia55otOmsXN5NhoudxdvtVly+hF3CLKDQ6E8tBYkFAeqxxus4EgAB4wCoWNgQeQAKyQGFWj2Vao1MrY-AZmrlakyAGstBB3nl3NxbHBytM5UI1MqAGSDGlEGU2p7oF6Yso3da0JBKw5c1ypZHZXLS2VQSPYmDjACqKvVmota1aZODGbUWfKlWhQgWcyWrHIrjUyBD8s6edWpKGSBGY1pVZubIbHLgOwTxzVkFgcFlmBQ6RkWEyqbyAEEwGBWj3uBSB9ThOMhTTxm2M0PKlb20yMVdq7d6o1mkhMKyjP5TDJ2CI3hAROOJ7eGe1qZvew7lnKt6XFiD64vcr7vn4JiBD+f4AROXK7EB4ruEAA).
Could anyone take a peek at that and tell me if I'm missing something ?

[a comment by an automated account is not shown]

**Comment by eps1lon** [comment]
Can you create a TypeScript playground (https://www.typescriptlang.org/play) instead? CodeSandbox is not very good at debugging TS issues.

**Reply by Mintoo200** [reply]
I'll see what I can do, I have no idea how Typescript Playground works or how to include react types and JSX in it, but I'll try my best.

**Reply by Mintoo200** [reply]
Wow, that was actually less painful than I expected, here is the [same minimal example in a Typescript playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgcilQ3wG4AocmATzCTgAUcwBnOAXnLjgB84BvLt2EAiFFGAoAtABsUAIyQyRALjgsYEgHYBzCsNHjJshUplIAJvOoiA-Go3a9Q7gF8hfQQe5iJ0uYrK9uqawLr63r7GAWaW1qohThFuFOSYAK5aGMAQWnAAstQAwriQWkhaMADqwDAAFhDpMMiYABRgzCxqTBCsAJQCQkQw6VB5ADwWwABuAgB0Cx29LK4AfAAqSBrjAPRT06sU7uRouRoFxaW5FdW1dS0ciMQwc5jQAO7iFi3jABLr+QAMgARGYAUXMIBuABpGJ1Vq0hBksjAcnlCiVwNdKjV6i12p1YURMAMvMJhqM8ojvHBJjM4MT2Pxia55otOmsXN5NhoudxdvtVly+hF3CLKDQ6E8tBYkFAeqxxus4EgAB4wCoWNgQeQAKyQGFWj2Vao1MrY-AZmrlakyAGstBB3nl3NxbHBytM5UI1MqAGSDGlEGU2p7oF6Yso3da0JBKw5c1ypZHZXLS2VQSPYmDjACqKvVmota1aZODGbUWfKlWhQgWcyWrHIrjUyBD8s6edWpKGSBGY1pVZubIbHLgOwTxzVkFgcFlmBQ6RkWEyqbyAEEwGBWj3uBSB9ThOMhTTxm2M0PKlb20yMVdq7d6o1mkhMKyjP5TDJ2CI3hAROOJ7eGe1qZvew7lnKt6XFiD64vcr7vn4JiBD+f4AROXK7EB4ruEAA)

**Comment by Mintoo200** [comment]
Just an update here if it helps. We though we found a workaround by declaring the discriminating union as a regular union : 
```ts
type Props = {
	'aria-label': string,
} | {
	'aria-labelledby': string,
} | {
	'aria-label': string,
	'aria-labelledby': string,
};
```
And it seemed to work at first but now we can't access `aria-label` (or `aria-labelledby` for that matter) in Props ...
Here's a [Typescript playground to illustrate](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgIilQ3wG4AoUSWAgOgHoBnGATwBslHa1HGzzyrMEjgAFHGEZwAvOThwAPnADec+evwoowFAFo2KAEZI2+AFxxm2gHYBzCurgBfNUtWP5m7XoPG2HABNDFnNLGBt7NXkXeTcojS0dfSMTUKtgOwcPLyTfE0DgtPCMyMcnCkEWYTgAQW8AGRS2GTEJRgBtAHJEnybOgF0BTABXawxgCGs4AFkWAGFcSGskaxgAdWAYAAsIYZhkTAAKMDaLcQhJAEoVNSIYYagpw-iAHgDgADc4HuS-aWUThcOjlen58P0nCpaNDAZInAA+eLyAAqnBgr3o7w+iPklwoLnIaEmzBm80WkxW602WwOLWQ6BgtEw0AA7loAgcXgAJZHTeoAEU+AFEOCBKQAaVpA+HPeQjMYwCZTWYLcAU1YbbYHY5tSVETDXdzqO4PJ5IuBvT5wfX-fWQ5TQ2iwxgI80otHml6Yz449R4tROcXkf2VarIawBJBQc6SF7IuBIAAeMBWASkEEMACskBh4S140mUxGpMpramoxZRgBrawQFlTGJwAD8cGWHyjags8YAZDcPGWIxXEMRGaqlpTkVUkHH4VlohV5eNJsPB1Ax+qYC8AKoJ5OpksIw5Goirizr5arIPyR3O8hOCzhyPRtrb+GG25Ie6PC3nylQmFtJC9CzneAhJtQ8CRpgKDDGwWCjEuUw1GAYCHO+8gmt+srqC8vqOC8j5Rr+qwDk+-wquSF5Utsuz7EgmCQj8eRsNI+DMhA+BwMBnqEWulF-ieZHKBRapUZqNL0Yx3i-CYrHsZx3H4YpcD+i4QA)
